### PR TITLE
[WIP] Add PReLU and LeakyReLU

### LINKED ISSUE #287: Add `tensor_ops::leaky_relu` and `nn::LeakyReLU`
See pytorch docs https://pytorch.org/docs/stable/generated/torch.nn.LeakyReLU.html#torch.nn.LeakyReLU.

Will need to figure out what the derivative of this function is, but that's probably online somewhere.

See https://github.com/coreylowman/dfdx/pull/397 for all the pieces you'll need to add for this.

### LINKED ISSUE #389: Add tensor_ops::prelu and nn::PReLU
Lately I've tried porting various computer vision networks to dfdx, and one function that seems quite popular in that field but isn't supported yet is PReLU. For now I just stay away from networks that use it, but it would be nice to have it available in the future.

https://pytorch.org/docs/stable/generated/torch.nn.PReLU.html#torch.nn.PReLU

ParametricReLU is bascially a LeakyReLU where the negative slope is a learned parameter, so there is going to be some overlap with the LeakyReLU implemenation (#287).

### PR DESCRIPTION
**Pull request by opfromthestart**
Fixes #287, #389, #584

Will add PReLU and LeakyReLU operations and modules.

- [x] Do NoneTape forward
- [x] Do OwnedTape forward
- [x] Do backward for LeakyReLU
- [x] Do backward for PReLU
- [x] (maybe later) do CUDA version

### CODE REVIEW

**Comment by opfromthestart** [comment]
@coreylowman I believe this is done for the most part.

**Comment by opfromthestart** [comment]
Nevermind it doesn't run.

File: src/tensor_ops/prelu/cpu_kernel.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        rhs: &Tensor<(), E, Cpu>,
+        rhs_grad: &mut <Self as crate::prelude::storage_traits::DeviceStorage>::Vec<Self::Elem>,
+        grad: &<Self as crate::prelude::storage_traits::DeviceStorage>::Vec<Self::Elem>,
+    ) -> Result<(), <Self::Output as HasErr>::Err> {
+        let op = PReLUKernelOp;
+        let rg = rhs_grad.get_mut(0).unwrap(); // TODO not unwrap
+        let r = *rhs.as_vec().get(0).unwrap();
+
+        // Should I do this?
+        let scale = E::from_f32(1.0 / lhs_grad.len() as f32).unwrap();
```
[the remark this answers is not in the document]
Comment (nkoppel) [reply src/tensor_ops/prelu/cpu_kernel.rs:80]: You shouldn't. Backward operations should always do the following:
```
input_grad_arg += partial_derivative_arg * output_grad
```

File: src/tensor_ops/prelu/prelu.cu [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    unsigned int i = blockIdx.x * blockDim.x + threadIdx.x;
+    float size_f = (float)size;
+    size_f = 1/size_f;
+
+    if (lhs[i] >= 0) {
+        lhs_grad[i] += out_grad[i];
+    }
+    else {
+        lhs_grad[i] += rhs * out_grad[i];
+        *rhs_grad += lhs[i] * size_f * out_grad[i];
```
Comment (nkoppel) [inline src/tensor_ops/prelu/prelu.cu:29]: This is unsound, because several threads could be trying to add to this location at once, which can cause a race condition, and for the sum to "lose" some of the inputs. Use atomicAdd when you need to have several threads modify the same location.

File: src/tensor_ops/prelu/cpu_kernel.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+impl<E: Float> BinaryDerivative<E> for PReLUKernelOp {
+    fn f(&self, x: &E, y: &E) -> E {
+        let zero = E::from(0.0).unwrap();
+        x.max(zero) + *y * x.min(zero)
+    }
+
+    fn dfdx(&self, x: &E, y: &E) -> E {
+        let zero = E::from(0.0).unwrap();
+        let one = E::from(1.0).unwrap();
```
Comment (nkoppel) [inline src/tensor_ops/prelu/cpu_kernel.rs:19]: This implementation is good, and can be directly used to implement the cpu binary kernels. (see next comment)

**Review by nkoppel — CHANGES_REQUESTED** [review]
I appreciate the effort that went in to implementing this, but I don't feel that this implementation makes good use of dfdx's internal features for implementing binary tensor operations. I recommend refactoring this and modeling it after the implementation of 'add', with LeakyReLU corresponding to ScalarAdd, and PReLU corresponding to BinaryAdd. Implementing these operations in this way will also be much simpler, as the high-level kernel logic is already implemented.

Using this version of the PReLU op, I think we should have PReLU0D and PReLU1D modules in `nn` to mirror PyTorch's interface. These would work by broadcasting their `a` fields to the shape of the input tensor before calling `prelu` on them.

**Comment by opfromthestart** [comment]
I did this using broadcast, but wont this have the same issue with concurrent overlaps?

File: src/nn/activations.rs [hunk 4]
```diff
@@ -55,9 +63,175 @@ where
...
+
+            fn try_forward(&self, input: Tensor<($($InDims),*), E, D, T>) -> Result<Self::Output, D::Err> {
+                input.try_prelu(self.a.clone().broadcast())
+            }
+        }
+    };
+}
+
+prelu1d!((B, C), Axis<0>);
+prelu1d!((B, C, M), Axes2<0, 2>);
```
Comment (nkoppel) [inline src/nn/activations.rs:213]: Is this how we should handle the 3d case? I would think this use case would usually occur when dealing with convolutions, in which case this should look like the following to mirror bias2d.
```suggestion
prelu1d!((C, B, M), Axes2<1, 2>);
```

**Comment by nkoppel** [comment]
> I did this using broadcast, but wont this have the same issue with concurrent overlaps?

It won't, because the current implementation of the backward pass of binary operations uses chunk_sum, which avoids data races by synchronizing the reads and writes of all threads.

The current version looks really good, nice job! Only things that need doing besides my previous comment are fixing the workflow errors and considering moving prelu from `activations` into its own file.

**Review by coreylowman — COMMENTED** [review]
Oh apparently I never actually clicked submit on my review - woops

File: src/nn/activations.rs [hunk 5]
```diff
@@ -55,9 +60,99 @@ where
...
+impl<Ax: Axes, S, E: Dtype, D: Device<E>, T: Tape<E, D>> Module<Tensor<S, E, D, T>> for PReLU<E, D>
+where
+    S: Shape<LastAxis = Ax> + ReduceShape<Ax>,
+    D: PReLUKernel<Tensor<S, E, D>, Tensor<(), E, D>, Output = Tensor<S, E, D>, Elem = E>,
+{
+    type Output = Tensor<S, E, D, T>;
+    type Error = D::Err;
+
+    fn try_forward(&self, input: Tensor<S, E, D, T>) -> Result<Self::Output, D::Err> {
+        input.try_prelu(self.a.clone())
```
[the remark this answers is not in the document]
Comment (nkoppel) [reply src/nn/activations.rs]: Until we have operator fusion, we should avoid doing this, because of the added memory use and overhead this brings.

[the remark this answers is not in the document]
Comment (coreylowman) [reply src/nn/activations.rs]: TBH I'd prefer simpler code that can be optimized later with operator fusion. Plus once we add operator fusion we don't have to go back and remove all this extra stuff. The less specialized kernels we have to write the better IMO. I think PReLU is niche enough where its not necessarily on the hot path.

[the remark this answers is not in the document]
Comment (nkoppel) [reply src/nn/activations.rs]: I'm not really convinced that the short term performance loss is worth it being convenient for a long-term optimization, but if we are going to do it for this operator, here's a more efficient implementation:
```rust
let scale = self.a.retaped::<T>().broadcast_like(t.shape());
let scaled = input.with_empty_tape() * scale;
(input.scalar_lt(0.0)).choose(scaled, input) 
```

[the remark this answers is not in the document]
Comment (coreylowman) [reply src/nn/activations.rs]: I'd like to start doing this more and more because:
1. It lowers the bar to contributing tensor ops because you don't have to be a cuda expert or even write multiple kernels
    1. As we look to adding additional devices (e.g. #597), each additional tensor op with custom kernel makes this harder.
2. Its more maintainable
3. It gives us a wider set of test cases for #607, which is becoming higher priority in my mind as I've look at optimizations recently

File: src/nn/activations.rs [hunk 6]
```diff
@@ -55,9 +60,99 @@ where
...
+#[derive(Default, Debug, Clone, Copy)]
+pub struct LeakyReLU<E: Dtype>(E);
+
+impl<E: Dtype> ZeroSizedModule for LeakyReLU<E> {}
+impl<E: Dtype> NonMutableModule for LeakyReLU<E> {}
+
+impl<Ax: Axes, S, E: Dtype, D: Device<E>, T: Tape<E, D>> Module<Tensor<S, E, D, T>> for LeakyReLU<E>
+where
+    S: Shape<LastAxis = Ax> + ReduceShape<Ax>,
+    D: PReLUKernel<Tensor<S, E, D>, Tensor<(), E, D>, Output = Tensor<S, E, D>, Elem = E>,
```
[the remark this answers is not in the document]
Comment (opfromthestart) [reply src/nn/activations.rs]: This should be done

File: src/tensor_ops/prelu/cpu_kernel.rs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        if x > &zero {
+            one
```
[the remark this answers is not in the document]
Comment (opfromthestart) [reply src/tensor_ops/prelu/cpu_kernel.rs]: This should be done

**Comment by nkoppel** [comment]
@coreylowman

**Comment by coreylowman** [comment]
@opfromthestart let's change it to non-device specific kernel

**Comment by opfromthestart** [comment]
@coreylowman I removed the device specific kernels, using nkoppel's implementation, and everything works

File: src/nn/activations.rs [hunk 8]
```diff
@@ -55,9 +60,175 @@ where
...
+}
+
+impl<C: ConstDim, E: Dtype, D: Device<E>> Default for PReLU1D<C, E, D> {
+    fn default() -> Self {
+        let dev = D::default();
+        Self {
+            a: dev.tensor(E::from_f32(0.25).unwrap()).broadcast(),
+        }
+    }
+}
```
Comment (coreylowman) [inline src/nn/activations.rs:169]: Let's remove this to make it consistent with the other nn modules (that you have to go through BuildModule)

File: src/nn/activations.rs [hunk 9]
```diff
@@ -55,9 +60,175 @@ where
     }
 }
 
+/// Calls [prelu()] with constant value.
+#[derive(Debug, Clone, Copy)]
+pub struct LeakyReLU<E: Dtype>(E);
```
Comment (coreylowman) [inline src/nn/activations.rs:65]:
```suggestion
pub struct LeakyReLU<E: Dtype>(pub E);
```

File: src/nn/activations.rs [hunk 10]
```diff
@@ -55,9 +60,175 @@ where
...
+}
+
+impl<E: Dtype, D: Device<E>> From<E> for PReLU<E, D> {
+    fn from(value: E) -> Self {
+        let dev = D::default();
+        Self {
+            a: dev.tensor(value),
+        }
+    }
+}
```
Comment (coreylowman) [inline src/nn/activations.rs:126]: Let's remove these so it's consistent with other nn modules - so it has to go through BuildModule to get access to device

**Review by coreylowman — CHANGES_REQUESTED** [review]
Looking great! I think after next round we can merge 👍

File: src/tensor_ops/prelu/mod.rs [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+impl<S: Shape, E: Dtype + Zero, D: Device<E> + ScalarCmpKernel<LtKernelOp, E>, T: Tape<E, D>>
+    TryPReLU<E> for Tensor<S, E, D, T>
+{
+    type Output = Self;
+
+    /// See [prelu]
+    fn try_prelu(self, rhs: E) -> Result<Self, Self::Err> {
+        let dev = D::default();
```
[the remark this answers is not in the document]
Comment (opfromthestart) [reply src/tensor_ops/prelu/mod.rs:86]: Done

File: src/nn/activations.rs [hunk 12]
```diff
@@ -55,9 +60,175 @@ where
...
+    (($($InDims:tt),*), $Axes:ty) => {
+        impl<E: Dtype, D: Device<E>, T: Tape<E, D> + Merge<NoneTape>, $($InDims: ConstDim),*> Module<Tensor<($($InDims),*), E, D, T>> for PReLU1D<C,E, D>
+        where ($($InDims),*): ReduceShapeTo<(C,), $Axes>,
+        Tensor<($($InDims),*), E, D, T>: TryPReLU<Tensor<($($InDims),*), E, D, NoneTape>, Output = Tensor<($($InDims),*), E, D, T>>,
+        {
+            type Output = Tensor<($($InDims),*), E, D, T>;
+            type Error = <Tensor<($($InDims),*), E, D, T> as HasErr>::Err;
+
+            fn try_forward(&self, input: Tensor<($($InDims),*), E, D, T>) -> Result<Self::Output, Self::Error> {
+                input.try_prelu(self.a.clone().broadcast())
```
[the remark this answers is not in the document]
Comment (opfromthestart) [reply src/nn/activations.rs:220]: done

File: src/nn/activations.rs [hunk 13]
```diff
@@ -55,9 +60,175 @@ where
...
+impl<S: ConstShape, E: Dtype, D: Device<E>, T: Tape<E, D>> Module<Tensor<S, E, D, T>>
+    for PReLU<E, D>
+where
+    Tensor<S, E, D, T>: TryPReLU<Tensor<S, E, D, NoneTape>, Output = Tensor<S, E, D, T>>,
+{
+    type Output = Tensor<S, E, D, T>;
+    type Error = <Tensor<S, E, D, T> as HasErr>::Err;
+
+    fn try_forward(&self, input: Tensor<S, E, D, T>) -> Result<Self::Output, Self::Error> {
+        input.try_prelu(self.a.clone().broadcast())
```
[the remark this answers is not in the document]
Comment (opfromthestart) [reply src/nn/activations.rs:146]: done

File: src/nn/activations.rs [hunk 14]
```diff
@@ -55,9 +60,175 @@ where
...
+    type Built = PReLU<E, D>;
+    fn try_build_on_device(device: &D) -> Result<Self::Built, <D>::Err> {
+        Self::Built::try_build(device)
+    }
+}
+
+/// Calls [prelu()] with learnable value.
+#[derive(Debug, Clone)]
+pub struct PReLU<E: Dtype, D: Device<E>> {
+    a: Tensor<(), E, D>,
```
[the remark this answers is not in the document]
Comment (opfromthestart) [reply src/nn/activations.rs:114]: done

File: src/tensor_ops/prelu/mod.rs [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+
+#[repr(C)]
+#[derive(Debug, Default, Clone, Copy)]
+pub struct PReLUKernelOp;
+
+#[repr(C)]
+#[derive(Debug, Default, Clone, Copy)]
+pub struct LeakyReLUKernelOp<E> {
+    slope: E,
+}
```
[the remark this answers is not in the document]
Comment (opfromthestart) [reply src/tensor_ops/prelu/mod.rs:20]: done

File: src/nn/activations.rs [hunk 16]
```diff
@@ -55,9 +60,175 @@ where
...
+impl<S: ConstShape, E: Dtype, D: Device<E>, T: Tape<E, D>> Module<Tensor<S, E, D, T>>
+    for LeakyReLU<E>
+where
+    Tensor<S, E, D, T>: TryPReLU<E, Output = Tensor<S, E, D, T>>,
+{
+    type Output = Tensor<S, E, D, T>;
+    type Error = D::Err;
+
+    fn try_forward(&self, input: Tensor<S, E, D, T>) -> Result<Self::Output, D::Err> {
+        Ok(leakyrelu(input, self.0))
```
[the remark this answers is not in the document]
Comment (opfromthestart) [reply src/nn/activations.rs:85]: done

File: src/tensor_ops/mod.rs [hunk 17]
```diff
@@ -233,6 +234,7 @@ pub use negate::negate;
 pub use normalize::normalize;
 pub use permute_to::PermuteTo;
 pub use pow::{powf, powi};
+pub use prelu::{leakyrelu, prelu, LeakyReLUKernelOp, PReLUKernelOp, TryPReLU};
```
[the remark this answers is not in the document]
Comment (opfromthestart) [reply src/tensor_ops/mod.rs:237]: done

File: src/nn/activations.rs (hunk 8 above)
Comment (opfromthestart) [reply src/nn/activations.rs:186]: done

File: src/nn/activations.rs (hunk 9 above)
Comment (opfromthestart) [reply src/nn/activations.rs:65]: done

File: src/nn/activations.rs (hunk 10 above)
Comment (opfromthestart) [reply src/nn/activations.rs:133]: done

File: src/tensor_ops/prelu/mod.rs [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+pub trait TryPReLU<T = Self>: HasErr {
+    type Output;
+
+    fn try_prelu(self, rhs: T) -> Result<Self::Output, Self::Err>;
+
+    fn prelu(self, rhs: T) -> Self::Output {
+        self.try_prelu(rhs).unwrap()
+    }
```
[the remark this answers is not in the document]
Comment (opfromthestart) [reply src/tensor_ops/prelu/mod.rs]: This did not work, I'm guessing that it is because the bound of ScalarCmpKernel<LtKernelOp, E> is not part of Device<E>. I did remove the assoc trait Output.

**Review by coreylowman — APPROVED** [review]
Awesome, thanks for all the work on this! 🚀
